src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -175,11 +175,257 @@ private BoundExpression BindSwitchExpression(SwitchExpressionSyntax node, Bindin
+            }
+            else if (TryFindIndexerOrIndexerPattern(node, inputType, argIsIndex: true, out indexerAccess, out Symbol? patternSymbol, diagnostics))
+            {
+                if (patternSymbol is PropertySymbol indexer)
+                {
+                    indexerSymbol = indexer;
+                    elementType = indexer.Type;
+                }
+                else
+                {
+                    Debug.Assert(indexerAccess is not null);
+                    elementType = indexerAccess.Type;
+                }
+            }
+            else
+            {
+                // PROTOTYPE(list-patterns) Cover any other type that we intend to cover, multi-dimensional arrays?
+                if (!hasErrors && !inputType.IsErrorType())
+                    Error(diagnostics, ErrorCode.ERR_UnsupportedTypeForListPattern, node, inputType);
+                hasErrors = true;
+                elementType = CreateErrorType();
+            }
+
+            ImmutableArray<BoundPattern> subpatterns = BindListPatternSubpatterns(node.Subpatterns, inputType, elementType, permitDesignations, ref hasErrors, out bool sawSlice, diagnostics);
+            return new BoundListPatternClause(node, subpatterns, sawSlice, indexerAccess, indexerSymbol, hasErrors);
+        }
+
+        private bool TryFindIndexerOrIndexerPattern(
+            SyntaxNode syntax, TypeSymbol receiverType, bool argIsIndex,
+            out BoundIndexerAccess? indexerAccess,
+            out Symbol? patternSymbol,
+            BindingDiagnosticBag diagnostics)
+        {
+            indexerAccess = null;
+            patternSymbol = null;
+            TypeSymbol argType = Compilation.GetWellKnownType(argIsIndex ? WellKnownType.System_Index : WellKnownType.System_Range);
+            if (argType.IsErrorType())
+            {
+                return false;
+            }
+
+            CompoundUseSiteInfo<AssemblySymbol> useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+            var lookupResult = LookupResult.GetInstance();
+            if (PerformPatternIndexerLookup(syntax, receiverType, lookupResult, out indexerAccess, argType, diagnostics, ref useSiteInfo))
+            {
+                _ = GetWellKnownTypeMember(argIsIndex ? WellKnownMember.System_Index__ctor : WellKnownMember.System_Range__ctor, diagnostics, syntax: syntax);
+                diagnostics.Add(syntax, useSiteInfo);
+                lookupResult.Free();
+                return true;
+            }
+
+            if (TryFindIndexOrRangeIndexerPattern(syntax, lookupResult, receiverOpt: null, receiverType, argIsIndex, out patternSymbol, diagnostics, ref useSiteInfo))
+            {
+                diagnostics.Add(syntax, useSiteInfo);
+                lookupResult.Free();
+                return true;
+            }
+
+            lookupResult.Free();
+            return false;
+        }
+
+        private bool PerformPatternIndexerLookup(
```

Comment:
>PerformPatternIndexerLookup [](http://example.com/codeflow?start=21&length=27)

Nit: `TryPerform`

Reply:
This is consistent with other namings. We don't "try" to perform the lookup, we actually do it. (would be true if it was GetPatternIndexer or sth)

Reply:
Changed my mind. Done.